plugins/basic/python/tk_maya_basic/plugin_logic.py

```diff
@@ -75,60 +79,134 @@ def login_user():
     Logs in the user to Shotgun and starts the engine.
     """
 
-    # Needed global to re-import the toolkit core later.
-    global sgtk
+    # Needed global to create the main progress bar.
+    global main_progress_bar
 
     try:
-        # When the user is not yet authenticated, pop up the Shotgun login dialog to get the user's credentials,
+        # When the user is not yet authenticated,
+        # pop up the Shotgun login dialog to get the user's credentials,
         # otherwise, get the cached user's credentials.
         user = sgtk.authentication.ShotgunAuthenticator().get_user()
 
     except sgtk.authentication.AuthenticationCancelled:
-        # When the user cancelled the Shotgun login dialog, keep around the displayed login menu.
+        # When the user cancelled the Shotgun login dialog,
+        # keep around the displayed login menu.
         standalone_logger.info("Shotgun login was cancelled by the user.")
         return
 
-    try:
+    # Get rid of the displayed login menu since the engine menu will take over.
+    delete_login_menu()
+
+    # Create a non-interruptable main progress bar (normally in the Help Line).
+    main_progress_bar = pm.ui.MainProgressBar(minValue=0, maxValue=100, interruptable=False)
+    main_progress_bar.beginProgress()
+
+    # Before bootstrapping the engine for the first time around,
+    # the toolkit manager may swap the toolkit core to its latest version.
+    plugin_engine.bootstrap(user,
+                            progress_callback=handle_bootstrap_progress,
+                            completed_callback=handle_bootstrap_completed,
+                            failed_callback=handle_bootstrap_failed)
+
+
+def handle_bootstrap_progress(step_number, message, current_index, maximum_index):
+    """
+    Callback function that reports back on the toolkit and engine bootstrap progress.
+
+    This function is executed in the main thread by the main event loop.
+
+    :param step_number: Current progress step number,
+                        from 1 to ``sgtk.bootstrap.ToolkitManager.MAX_PROGRESS_STEP_NUMBER``.
+    :param message: Progress message to report.
+    :param current_index: Optional current item number being looped over.
+    :param maximum_index: Optional maximum item number being looped over.
+    """
+
+    if maximum_index and maximum_index > 1:
+        message = "%s (%s of %s)" % (message, current_index+1, maximum_index+1)
+
+    message = "Bootstrapping %s: %s" % (manifest.engine_name, message)
```

Comment:
This is the list of messages that are displayed aside the main progress bar on Maya Help line as the bootstrap progresses:

> Bootstrapping tk-maya: Resolving project...
> Bootstrapping tk-maya: Resolving configuration...
> Bootstrapping tk-maya: Updating configuration...
> Bootstrapping tk-maya: Starting up Toolkit...
> Bootstrapping tk-maya: Downloading tk-maya v0.5.10 (1 of 5)...
> Bootstrapping tk-maya: Downloading tk-multi-shotgunpanel v1.2.6 (2 of 5)...
> Bootstrapping tk-maya: Downloading tk-multi-loader2 v1.11.2 (3 of 5)...
> Bootstrapping tk-maya: Downloading tk-framework-shotgunutils v4.3.1 (4 of 5)...
> Bootstrapping tk-maya: Downloading tk-framework-qtwidgets v2.2.1 (5 of 5)...
> Bootstrapping tk-maya: Resolving context...
> Bootstrapping tk-maya: Launching Engine...

Updating to say "Initializing Shotgun:".